### ISSUE TITLE
agent gets into bad state and drops metrics

### ISSUE THREAD

**Issue by nicklan** [grafana/agent#248]
We've noticed our agents can get into a state where they start dropping a majority of the metrics they scrape. Associated logs are:

```
ts=2020-11-06T19:49:08.929711874Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423202
ts=2020-11-06T19:49:08.929752731Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423203
ts=2020-11-06T19:49:08.929791189Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423204
ts=2020-11-06T19:49:08.929834035Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423205
ts=2020-11-06T19:49:08.929871078Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423206
ts=2020-11-06T19:49:08.929920881Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1423207
```

over and over and over (like 10s per second)

Restarting the agent immediately fixes the issue with no change in the targets or metrics it's scraping, so I do suspect an internal issue here.

Please let me know if there are additional details that would be helpful, thanks.

**Comment by rfratto** [comment]
Hi, sorry you're running into this; it looks like this would happen if the Agent were to incorrectly remove a series from the WAL while it still exists. 

Can you share your Agent config and the number of active series/sample throughput the Agent is seeing?

You can find the active series with `agent_wal_storage_active_series` and the sample throughput with the rate of `agent_wal_samples_appended_total`.

**Comment by nicklan** [comment]
Thanks for the reply.

agent_wal_storage_active_series is about 750k, which is pretty high i realize. agent_wal_samples_appended_total is about 25k/sec

the config is quite complex. Here's some of it, minus the actual scrape jobs:

```
prometheus:
  global:
    scrape_interval: 30s
    scrape_timeout: 20s
    external_labels:
      [snip]
  configs:
    - name: agent
      host_filter: false
      remote_write:
        - url: "http://localhost:9900/api/v1/prom/remote/write"
          remote_timeout: 30s
          queue_config:
            capacity: 10240 # default = 500
            max_shards: 200 # default = 1000
            min_shards: 10 # default = 1
            max_samples_per_send: 2000 # default = 100
            batch_send_deadline: 60s # default = 5s
            min_backoff: 100ms # default = 30ms
            max_backoff: 10s # default = 100ms
      scrape_configs:
...
```

Then we have a number of scrape configs. If there's something in particular that would be interesting there lmk.

**Comment by rfratto** [comment]
Thanks, nothing really seems to stand out there as problematic. 

Could you share the stats from `agentctl wal-stats`? In particular, I'm most interested validating that there's a non-zero number of invalid refs, but the entire first section (e.g., not the per-target stats table) should be helpful here. 

Do you see any log messages with `could not truncate WAL` in them? Other interesting error messages would be `truncating segments failed` and `delete old checkpoints`

**Comment by nicklan** [comment]
So, more info. It turned out that restarting didn't really fix the issue, just caused it to go away for a short while (note we don't keep the WAL around when we restart). I suspect this was caused simply by the service being overloaded. I've currently resolved it in our environment by giving the service more CPU, and reducing the number of series it has to scrape.

I think being overloaded causes the agent to get into a bad interaction with M3 (our remote write target) where it's trying to send stale metrics, and then "bad things happen".  here's some logs that might be helpful (note localhost is just a proxy loadbalancer that fowards the request on to m3):

```
[snip lots of "Dropped sample" errors]
ts=2020-11-07T16:24:39.859601981Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1282982
ts=2020-11-07T16:24:44.942256413Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=warn remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Error process checkpoint" err="readCheckpoint: getSegmentSize: stat /tmp/agent/data/5f32452cd0442efde5c553cc47441659/wal/checkpoint.00000313/00000001: no such file or directory"
level=warn ts=2020-11-07T16:25:07.843418443Z caller=logging.go:51 msg="GET /metrics 22.096759646s, error: write tcp 127.0.0.1:80->127.0.0.1:34166: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=info ts=2020-11-07T16:25:23.644767857Z caller=wal.go:378 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="series GC completed" duration=16.899220311s
level=info ts=2020-11-07T16:25:23.645599529Z caller=checkpoint.go:96 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="Creating checkpoint" from_segment=316 to_segment=317 mint=1604765589000
level=warn ts=2020-11-07T16:25:42.84559189Z caller=logging.go:51 msg="GET /metrics 27.59167437s, error: write tcp 127.0.0.1:80->127.0.0.1:45066: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:26:12.644649593Z caller=logging.go:51 msg="GET /metrics 27.302112709s, error: write tcp 127.0.0.1:80->127.0.0.1:46604: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:26:35.399180513Z caller=logging.go:51 msg="GET /metrics 31.857604455s, error: write tcp 10.2.253.131:80->10.2.218.125:60696: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/2.19.2; X-Prometheus-Scrape-Timeout-Seconds: 25.000000; "
level=warn ts=2020-11-07T16:26:50.348707604Z caller=logging.go:51 msg="GET /metrics 35.006709825s, error: write tcp 127.0.0.1:80->127.0.0.1:48268: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:27:10.544739691Z caller=logging.go:51 msg="GET /metrics 25.102837689s, error: write tcp 127.0.0.1:80->127.0.0.1:50272: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:27:45.351413099Z caller=logging.go:51 msg="GET /metrics 29.907950798s, error: write tcp 127.0.0.1:80->127.0.0.1:51790: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=info ts=2020-11-07T16:27:58.250369288Z caller=wal.go:443 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="WAL checkpoint complete" first=316 last=317 duration=2m51.504820564s
level=warn ts=2020-11-07T16:28:46.245615631Z caller=logging.go:51 msg="GET /metrics 31.000028605s, error: write tcp 127.0.0.1:80->127.0.0.1:53234: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:28:46.252218146Z caller=logging.go:51 msg="GET /metrics 38.510583901s, error: write tcp 10.2.253.131:80->10.2.218.128:56098: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/2.19.2; X-Prometheus-Scrape-Timeout-Seconds: 25.000000; "
level=warn ts=2020-11-07T16:28:46.341735338Z caller=logging.go:51 msg="GET /metrics 38.800098824s, error: write tcp 10.2.253.131:80->10.2.126.81:37850: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=info ts=2020-11-07T16:29:39.743421137Z caller=wal.go:378 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="series GC completed" duration=15.689756638s
level=warn ts=2020-11-07T16:29:40.745826267Z caller=logging.go:51 msg="GET /metrics 33.002319517s, error: write tcp 10.2.253.131:80->10.2.218.128:39326: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/2.19.2; X-Prometheus-Scrape-Timeout-Seconds: 25.000000; "
level=warn ts=2020-11-07T16:29:40.747983502Z caller=logging.go:51 msg="GET /metrics 33.20617642s, error: write tcp 10.2.253.131:80->10.2.126.81:51686: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=error ts=2020-11-07T16:29:40.752951868Z caller=http.go:99 agent=prometheus msg="failed to write response" err="write tcp 10.2.253.131:80->10.2.89.98:41450: i/o timeout"
level=warn ts=2020-11-07T16:29:40.753061619Z caller=logging.go:51 msg="GET /agent/api/v1/targets 40.511481624s, error: write tcp 10.2.253.131:80->10.2.89.98:41450: i/o timeout ws: false; Accept-Encoding: identity; Connection: close; User-Agent: Python-urllib/3.6; "
level=warn ts=2020-11-07T16:29:40.847446977Z caller=logging.go:51 msg="GET /metrics 25.504398354s, error: write tcp 127.0.0.1:80->127.0.0.1:56298: write: broken pipe ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=info ts=2020-11-07T16:31:10.549309161Z caller=wal.go:378 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="series GC completed" duration=15.804577535s
level=info ts=2020-11-07T16:31:10.549967801Z caller=checkpoint.go:96 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="Creating checkpoint" from_segment=318 to_segment=319 mint=1604765782000
ts=2020-11-07T16:31:15.241864954Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=error remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="non-recoverable error" count=2000 err="server returned HTTP status 400 Bad Request: {\"error\":\"bad_request_errors: count=1, last=failed to meet consistency level majority with 0/3 success, 3 nodes responded, errors: [error writing to host {\\\"name\\\":\\\"m3db-rep0-4\\\"}: Error({Type:BAD_REQUEST Message:datapoint too far in past: id={__name__=\\\""
[snip lots more BAD_REQUEST errors]
level=info ts=2020-11-07T16:55:52.06055664Z caller=wal.go:443 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="WAL checkpoint complete" first=326 last=327 duration=2m46.918098564s
level=warn ts=2020-11-07T16:56:08.742334958Z caller=logging.go:51 msg="GET /metrics 31.000723085s, error: write tcp 10.2.253.131:80->10.2.218.128:55032: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/2.19.2; X-Prometheus-Scrape-Timeout-Seconds: 25.000000; "
level=warn ts=2020-11-07T16:56:08.743946839Z caller=logging.go:51 msg="GET /metrics 31.202188394s, error: write tcp 10.2.253.131:80->10.2.126.81:52242: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=warn ts=2020-11-07T16:56:19.255426273Z caller=logging.go:51 msg="GET /metrics 34.008069426s, error: write tcp 127.0.0.1:80->127.0.0.1:45678: i/o timeout ws: false; Accept: application/openmetrics-text; version=0.0.1,text/plain;version=0.0.4;q=0.5,*/*;q=0.1; Accept-Encoding: gzip; User-Agent: Prometheus/v0.6.1; X-Prometheus-Scrape-Timeout-Seconds: 20.000000; "
level=info ts=2020-11-07T16:57:24.345229215Z caller=wal.go:378 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="series GC completed" duration=23.801920535s
level=info ts=2020-11-07T16:58:47.365599578Z caller=wal.go:378 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="series GC completed" duration=10.313332264s
level=info ts=2020-11-07T16:58:47.366287617Z caller=checkpoint.go:96 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 msg="Creating checkpoint" from_segment=328 to_segment=329 mint=1604767256000
ts=2020-11-07T17:00:07.242252148Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Done replaying WAL" duration=5m12.000705569s
ts=2020-11-07T17:00:09.342354001Z caller=dedupe.go:112 agent=prometheus instance=5f32452cd0442efde5c553cc47441659 component=remote level=info remote_name=5f3245-25d49b url=http://localhost:9900/api/v1/prom/remote/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=1295264
[rinse, repeat]
```

**Comment by rfratto** [comment]
Thanks for that, this is starting to click more now. Since you noticed increasing the available CPU made a difference, I suspect the Agent is too aggressive in deleting metrics from the WAL. 

It seems like it's possible for series to be deleted from the WAL in the middle of remote_write divvying out samples it has just read from the WAL. You're more likely to run into this with a smaller number of max_shards (as you have) and if CPU is a bottleneck (which you seem to have proven). The main issue is that the Agent deletes data from the WAL using the most recent timestamp successfully written to remote_write. However, one remote_write shard could be ahead of the others, like if one shard is retrying a batch after getting a 4xx, but the WAL deletion logic has no way of knowing this, and will incorrectly delete that data early. 

I think #138 is related and will help with you're seeing here; if the Agent guarantees data sits in the WAL for at least the configured `batch_send_deadline`, I don't think you'll run into this again.

**Comment by rfratto** [comment]
@nicklan If you're able to use non-release docker images, the master branch and image `grafana/agent:master-d6661ab` has #253 which might fix this problem.

**Comment by nicklan** [comment]
Thanks! Since we've resolved it for now with a bump in CPU, we can wait for this to land in a release.  Unfortunately the environment where this happens isn't one we can just test in and needs to be kept as stable as possible.

**Comment by rfratto** [comment]
Completely understandable :) I'm going to close this issue as fixed by #253 for now, but please let me know if you run into this again after the next release and I'll be happy to re-open.
